### DISCUSSION TITLE
How to point to compiled assets outside controllers/views

### DISCUSSION THREAD

**Discussion by afdev82** [shakacode/shakapacker#471] [Q&A] [answered]
Hi,

I am trying to include the compiled assets in my test environment (rspec inside a Ruby on Rails project).
To do that, I need the methods of the [shakapacker view helper](https://github.com/shakacode/shakapacker/blob/main/lib/shakapacker/helper.rb) available in my helper (eg. `asset_path`), to be able to point to the compiled assets and load them with MiniRacer ([this is why](https://github.com/rubyjs/mini_racer/discussions/301)).
Currently I am getting the following error:

```
NoMethodError:
         undefined method `asset_path' for an instance of Configuration
```

Is there a way or a better way to achive this?
Thank you for your help!

**Comment by tomdracz** [comment, marked as answer]
You can try include the relevant helpers in your class. Shakapacker relies on some Rails helpers so you would need them also. Something like:

```
class MyTestClass
  include Shakapacker::Helper
  include ActionView::Helpers
  include ActionView::Helpers::AssetTagHelper
end

MyTestClass.new.asset_pack_path('my_pack_path.js')
```

You can see how we do something similar (though with extend as we want the module methods to be class rather than instance methods) in one of the specs https://github.com/shakacode/shakapacker/blob/main/spec/shakapacker/helper_spec.rb

**Comment by afdev82** [comment]
I have found the path of the compiled assets, but in my tests running the JS with MiniRacer, I am getting the following error:

```
Failure/Error: ctx.load(bundle_path)
     
     MiniRacer::RuntimeError:
       ReferenceError: self is not defined
     # JavaScript at <anonymous>:1:1

```
I don't know if it's an error of MiniRacer or I have to include something more to be able to use the bundle.
I have opened the file, and in fact it starts with "self...":


```
(self["webpackChunkad_notam"] = self["webpackChunkad_notam"] || []).push([["configurator_ssr"],{
```

Are there any examples of how to use a webpack bundle in rspec tests?
I have added now the line `var self = globalThis` first and then loaded the bundle, the error is gone... But I am not sure it's the right way to solve it, because I am getting other errors of not defined variables (eg. I18n), that are defined in the bundle.

**Reply by tomdracz** [reply]
That would suggest potentially trying to execute code that relies on some Web APIs and/or `window` object in node-like environment.

To get around error itself, one could set https://webpack.js.org/configuration/output/#outputglobalobject `globalObject` to `this` or `globalThis` in the webpack output configuration. This is what for example https://github.com/shakacode/shakapacker/blob/1a89919aac4e1d3dd65de1b041b8ad4607f10b36/spec/dummy/config/webpack/serverWebpackConfig.js#L48 SSR config in our repo does.

However if your code runs in non-browser context and uses some browser only APIs then that might cause other trouble.

Separately, can you elaborate on your use case trying to run the JS code through mini_racer? Curious on the intention here.

**Reply by afdev82** [reply]
Thank you for the reply,

I have seen that option, but I wanted to ask first, to know the best approach.
I will try to play around with the config and see if it solves the problem.

I am working on an online product configurator where the most of the logic to build the product is done in JS in the browser (both to collect all the components and to draw them in the canvas).
For that reason I have to run the same JS code (except the draw in the canvas side) server side to build the product and to be able to test it with rspec (eg. presence of a component that I expect to be there).
Before this change (move to webpack), I have used the standard rails asset pipeline for the JS in the browser and the `ctx.load` method of mini_racer server side to include the assets that I needed for the tests and it worked fine.
Now I thought it would be nice to include the assets using webpack in the tests too to have only one place to manage them.

**Reply by tomdracz** [reply]
I think using globalThis is probably the way to go then.

Thanks for the explanation of the setup. When you say running same code server size - is that in the deployed version or just for tests? My first instinct would be probably to offload JS tests into dedicated JS test runner rather than trying to squeeze it into RSpec. I tend to avoid trying to mix the two or cover JS client side behaviours with E2E system specs, as otherwise it's quite easy to get into those obscure issues!

**Reply by afdev82** [reply]
In fact it's more complex than that... I use the JS code also server side to have the data needed to generate PDFs, calculate prices, generate preview images, etc... So it's not only for the test suite.
I am still facing problems of not defined variables (the problem with self is solved), it's very strange...
If I include the following files in the following way in mini_racer (like it was before), it's finally working:

```
    ctx.load(Rails.root.join('app', 'javascript', 'configurator', 'core', 'compatibility_layer.js'))
    ctx.load(Rails.root.join('node_modules', 'inflection-js', 'inflection.js'))
    ctx.load(Rails.root.join('app', 'assets', 'javascripts', 'i18n', 'i18n.js'))
    ctx.load(Rails.root.join('app', 'assets', 'javascripts', 'i18n', 'translations.js'))

    bundle_path = Rails.root.join('public' + asset_pack_path('configurator_ssr.js'))
    ctx.load(bundle_path)
```

If I include the first 3 files in the pack:

```
import './configurator/core/compatibility_layer.js';
import 'inflection-js/inflection';
import '../assets/javascripts/i18n/i18n.js';
import '../assets/javascripts/i18n/translations.js';
```

I'm getting the following error:
```
MiniRacer::RuntimeError:
     #   ReferenceError: I18n is not defined
     #   JavaScript at eval (webpack://ad_notam/./app/assets/javascripts/i18n/translations.js?:1:1)

```
But `i18n` is defined in   '../assets/javascripts/i18n/i18n.js', before '../assets/javascripts/i18n/translations.js'.
The file `i18n.js` is part of the [i18n-js gem](https://github.com/fnando/i18n-js).

And if I comment the two files, I am getting other errors related to variables defined in './configurator/core/compatibility_layer.js'.
I am including on top of all a "compatibility_layer.js" file, that is responsible to define some variables that are needed in the other files (it's responsible also to handle $, window).
It seems the file don't get included somehow... But If I have a look at the packed file, it's there.

I have found a solution now (including some files using mini_racer), but it would be nice to understand why it's happening.

**Reply by tomdracz** [reply]
Thanks for the context. Hard to say much more without seeing the actual code.  I'm not sure about the import execution order and what determines it, but from a quick test:

Given `another_test.js`:

```
testfoo = 'bar'
testbar = 'baz'
```

And `yet_another.js`:
```
console.log('testfoo', testfoo)
console.log('testbar', testbar)
```

```
import '../src/actions/another_test.js'
import '../src/actions/yet_another.js';
```

Will work and below will not


```
import '../src/actions/yet_another.js';
import '../src/actions/another_test.js'
```

That makes sense, execution order is retained. However if for example I would reference `import '../src/actions/yet_another.js';` in another pack. It won't be executed again, as it was already loaded. I don't know an awful lot about internal resolution and handling though. Stuff like code splitting, dependency tree resolution etc might influence that order.

If you're dealing with files defining some global variables and would want to ensure the load order, I would change those imports to requires as:
```
require('./configurator/core/compatibility_layer.js');
require('inflection-js/inflection');
require('../assets/javascripts/i18n/i18n.js');
require('../assets/javascripts/i18n/translations.js');
```

This will guarantee that load order will be as defined.

Elsewhere, if your `translations.js` depends on some things in `i18n.js` you would normally import it in `translations.js` to be specific about this dependency, similar to:

```
import { I18n } from '../assets/javascripts/i18n/i18n.js'
```

Obviously, this will only work if that JS file is actually having some exports. From what I can understand about your code it seems to just have bunch of global variables, functions etc defined in the classes rather than fully fledged modules imports and exports. In this case, `require` is probably your best bet

**Reply by afdev82** [reply]
It seems that in tests now it's working, with the specific webpack configuration used.
Anyhow, as I have explained before, I need to run the same code also in the production environment and if I do that, the test configuration is not used.
So, instead of a "test" configuration, in fact I need a configuration that is used when the code is run "server side" (not depending on the Rails environment).
Is it possible to achieve that?

**Reply by tomdracz** [reply]
In general, you could achieve that with multiple config objects https://webpack.js.org/configuration/configuration-types/#exporting-multiple-configurations if you provide array of config objects to webpack, it will execute all of them. This could be achieved to generate server specific code with distinct config.

You can see example of how we've implemented this in the dummy spec app here https://github.com/shakacode/shakapacker/tree/main/spec/dummy/config/webpack

**Reply by afdev82** [reply]
Yes, in my files I'm not using module exports, they are defining global objects to handle functions, etc... inside them (in fact there is a global object used as a namespace and inside of it all the logic / components needed.

I'll have a look at the references provided, thank you for your help!

**Reply by afdev82** [reply]
There is something I don't understand, the serverConfig is used if the ENV variable SERVER_BUNDLE_ONLY is set.
What I need is that if I load the bundle in mini_racer in this way:

```
bundle_path = Rails.root.join('public' + asset_pack_path('configurator_ssr.js'))
    ctx.load(bundle_path)
```
It gets compiled following the server specific webpack config.

I have changed the config for the test environment, because I use mini_racer also in the rspec tests and it was working.
But to choose the webpack config based on the RAILS_ENV it doesn't work in general, because the server side rendering is used also in the production environment.

Should I first precompile once the bundle used for the server side rendering and point to that file? In this case I could not use the `asset_pack_tag` at all, no hashes, chunks... In production it could work, in development I would need to use the shakapacker dev server to watch for changes and recompile the bundle.
Is it the way to go? I would need only to remember to precompile the server bundle every time I publish with a specific command (at the moment I use only the asset pipeline and all packs get compiled).

**Reply by tomdracz** [reply]
The linked is just an example. You don't have to add any conditionality. You could have single webpack config regardless of the environment and as long as it specifies two different configs, they would be applied.

**Reply by afdev82** [reply]
I think now I have got it.
I could have a webpack config only for the pack that I will use server side and another config that I use for all the other packs.

**Reply by afdev82** [reply]
It's working as expected.
Many thanks for your precious help! 🥇
